radareorg/cutter
# Refactored Decompiler Widget and R2Dec Plugin to use RAnnotatedCode

**Your checklist for this pull request**


**Detailed description**
This PR refactors decompiler widget and r2dec cutter plugin to use RAnnotatedCode instead of the custom AnnotatedCode we have been using. Using RAnnotatedCode will enable us to have more data about the decompiled code that is essential for a good decompiler.

The following are the major changes I have made.

1. Refactored `R2Dec::decompileAt()` to emit `RAnnotatedCode*`.
2. Deleted structures `AnnotatedCode`, `CodeAnnotation`, and functions that made to handle these structures.
3. Used a smart pointer to represent code in `DecompilerWidget`
`std::unique_ptr<RAnnotatedCode> code;`
4. Refactored slot `decompilationFinished(AnnotatedCode code)` to `decompilationFinished(RAnnotatedCode *code)`
5. Made two static functions `static ut64 offsetForPosition(RAnnotatedCode *codeDecompiled, size_t pos)` and `static size_t positionForOffset(RAnnotatedCode *codeDecompiled, ut64 offset)`. These are the corresponding functions that were deleted from `AnnotatedCode`. Made required changes in all functions that use these two functions.
 



**Test plan (required)**


Use only R2Dec decompiler for these tests, other decompilers are outside the scope of this PR.
1. Fetch the PR and compile it.
2. Open the decompiler widget and make sure that the correct decompiled code is being shown while you select offsets from functions and other widgets.
3. Make sure that the decompiler widget is seeking the correct address while clicking on functions(and other offsets) in the decompiler widget.
4. Make sure that invalid offsets are not causing Segmentation Fault.




**Notes**
Merge all these three PRs together to ensure smooth functioning of all decompilers.
Refactored decompiler widget and R2Dec plugin: [Cutter #2227 ](https://github.com/radareorg/cutter/pull/2227)
Refactored R2Ghidra plugin that emits RAnnotatedCode: [R2Ghidra-dec #112](https://github.com/radareorg/r2ghidra-dec/pull/112)
Refactored retdec plugin that emits RAnnotatedCode:[ Florian's fork #1](https://github.com/thestr4ng3r/retdec-r2plugin/pull/1), after this PR is verified, I will send it to retdec's repo.
